05-vue-router/05-AuthGuard/router/index.ts

```diff
@@ -0,0 +1,94 @@
+import { createRouter, createWebHistory, type RouteLocationNormalized, type Router } from 'vue-router';
+import { isAuthenticated } from '../services/authService.js';
+
+type TMiddlewareContext = {
+  from: RouteLocationNormalized;
+  to: RouteLocationNormalized;
+  router: Router;
+};
+
+type TMiddleware = (context: TMiddlewareContext) => Promise<void>;
+
+const log: TMiddleware = ({ from, to }) => {
+  console.log({ from: from.path, to: to.path });
+  return Promise.resolve();
+};
+
+const auth: TMiddleware = async ({ router, to }) => {
+  if (isAuthenticated()) {
+    return;
+  }
+
+  await router.push({ name: 'login', query: { ...to.query, from: to.path } });
+};
+
+const guest: TMiddleware = async ({ router }) => {
+  if (!isAuthenticated()) {
+    return;
+  }
+
+  await router.replace({ name: 'index' });
+};
```

Comment:
In my opinion, it's better to keep middlewares implementing the default interface of router hooks - `to, from` as params and returned value that can be directly used as value for VueRouter hook. 

Then we can use middleware as a general hook outside of our custom middleware and keep the interface is known to any Vue developer.

I know, that it is from the Markus article. Though, in new VueRouter version with async hooks it is easy to implement.